M2/Macaulay2/packages/SLPexpressions.m2

```diff
@@ -884,6 +889,28 @@ FY = value(diff(Y,F),h)
 assert ( value(compress diff(Y,G/F), h) == (GY*value(F,h) - value(G,h)*FY)/(value(F,h))^2 )
 ///
 
+
+------------------------
+-- expression, net, html
+expression InputGate := g -> expression g.Name
+expression SumGate := g -> Parenthesize sum(g.Inputs,expression)
+expression ProductGate := g -> Parenthesize product(g.Inputs,expression)
+expression DivideGate := g -> expression g.Inputs#0 / expression g.Inputs#1
+html GateMatrix := html @@ expression
+eGM:=
+expression GateMatrix := g -> new MatrixExpression from applyTable(g,expression)
+expression DetGate := g -> (expression det) eGM g.Inputs
+printLargeGate = g -> VerticalList { class g, "depth"=>depth g, "#gates"=> countGates g }
+net Gate := g -> (
+    n := net expression g;
+    if width n <= printWidth and height n <= printWidth then n
+    else net printLargeGate g
+    )
+html Gate := g -> (
+    e := expression g;
+    html if width net e <= printWidth then e else printLargeGate g
```

Comment:
the problem with testing `width net e` is that it forces to first produce the net output before deciding whether it's worth printing everything or not. It's usually a better idea to find a simpler criterion (number of ... whatever gates are made of) to make this choice. same remark for `net Gate`.

Reply:
This sounds reasonable. One attribute that is computed fast (faster than generating a net) is the depth. Printing is now attempted only if the depth is small.

Reply:
I would completely remove the part about `net e` in the tests, but I won't argue.